035-ca-abstraction.md

```diff
@@ -0,0 +1,315 @@
+```
+
+Then the "CA certificate trust state machine":
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_NEW // trusted everywhere it needs to be, but no EE certificates yet issued
+   |
+   v
+  TRUSTED_USED // trusted everywhere, EE certificates issued
+   |
+   v
+  PHASE_OUT // trusted everywhere, but being phased out
+   |
+   v
+  ZOMBIE // still trusted by some component, but not relied on by any party.
+```
+
+Now let's describe the processes.
+
+1. The issuing operator needs to figure out which EE certificates are missing (or in their renewal period, or otherwise need replacing, e.g. via annotation). This need for a certificate might exist independently of the end entity. For example, if we're going to scale up the Kafka cluster then we need a certificate for the new broker before the new broker exists. **The issuing operator puts the end entity in the `NEW` state.**
+2. (**See step 1 in the diagram.**) The issuing operator requests a new certificate. **The issuing operator puts the end entity in the `REQUESTED` state.**
+3. (**See step 2 in the diagram.**) The issuing operator cannot assume that issuance will happen quickly (i.e. that `pollForRequestedCertificate()` returns a non-empty `Optional` immediately), so it needs to be able persist the token obtained from `requestCertificate()` in a `Secret`, _and possibly complete the current reconciliation_ (i.e. issuance can be very asynchronous). 
+(An example of how this could happen is where a `CertificateIssuer` implementation uses cert-manager; `requestCertificate()` creates a cert-manager `CertificateRequest` CR but while the cert manager operator is not running no certificate will be issued).
+
+4. Once a certificate is issued, the issuing operator needs check whether the CA certificate is trusted. 
+   I.e. it needs to know what trust states all the existing CA certificates are in. 
+   If it is not trusted **the issuing operator puts the end entity in the `PENDING` state.** and **puts the CA certificate in the `UNTRUSTED` state**. (**See step 3 in the diagram.**)
+
+> From the above it is clear that there is a `Secret` which stores tokens and issued certificates and their current state, and that only the issuing operator needs access to it.
+> It's also clear that there is a `Secret` which stores CA certificates, and their current state, and which needs to be accessed by the issuing operator and the cluster operator.
+
+#### Process for ensuring CA trust
+
+When an EE certificate is issued via a `CertificateIssuer` it might have been signed by a CA that is not yet trusted.
+The cluster operator (for the cluster CA), or the user operator (for the clients CA) needs to detect this lack of established trust and ask the cluster operator to trust the new CA. 
+The cluster operator and the user operator run in separate processes, so this will be done via a shared `Secret`, for example:
```

Comment:
That assumes that the user knows beforehand which CA cert will be used by the issuer. Is that necessarily always true? They might know _today_, but what about some future time when that CA cert is close to expiring and the CA starts using a new one? If the person operating the cluster hasn't configured trust in that clients CA cert then the UO would happily issue user EE certs which won't work. Or should the CO pass the set of trusted certs to the UO, so the UO can check the issued certs will be trusted.

Reply:
It may be that it's a mistake to try to have any common abstraction that covers the cluster CA and the clients CA. From the PoV of the Kafka cluster it would work for the user to specify a set of CA certs to trust (and roll the brokers when that changed). And for the UO just make it a requirement that there is a CA (possibly intermediate) cert and key that can be given to the UO to issue certificates. In other words, not use this CA abstraction at all for the UO. That would certainly simplify a number of things, like the need for the shared Secret. At the cost of making the UO less broadly applicable, since it would still be making this assumption about being able to issue certificates itself.